### ISSUE TITLE
tsconfig: exclude not working as expected

### ISSUE THREAD

**Issue by bpasero** [Microsoft/TypeScript#4925]
I would like to exclude node_modules from my compilation but still be able to reference d.ts files within. Imagine the following scenario: I want to depend on one node module that comes with a d.ts file. If I exclude node_modules, this d.ts file will never be found. If I do not exclude node_modules, I might drag in other d.ts files that I dont want.

The best solution imho would be:
- allow me to exclude node_modules
- if there is a import for a module, still use node.js module resolution logic to find the d.ts file

In other words, never let the exclude property in tsconfig.json prevent the tsc look up logic to look into node_modules as needed.

**Comment by mhegazy** [comment]
What you are explaining is the expected behavior. if you seeing something different it is a bug. can you please provide more information and preferably a repro project?

**Comment by bpasero** [comment]
@mhegazy ok I might have spoken too soon, I can get tsc to compile the following project: https://github.com/bpasero/attachments/blob/master/test.zip?raw=true

What worries me is that I can rename the folder "issue" inside node_modules to "issue2" and everything still compiles. I can also rename issue.d.ts to issue2.d.ts and it will still work.

I would probably expect that if I do a `import * as issue from 'issue';` that you would only ever look for a issue.d.ts in a folder called issue anywhere in the node_modules hierarchy?

**Comment by bpasero** [comment]
@mhegazy ok and now I have a case where the exclude seems to fail: https://github.com/bpasero/attachments/blob/master/test_2_fail.zip?raw=true

The typescript module is within node_modules and even though I exclude node_modules in the tsconfig.json I get errors when I compile:

node_modules/typescript/lib/lib.core.d.ts(1168,28): error TS1005: ';' expected.
node_modules/typescript/lib/lib.core.d.ts(1168,31): error TS1005: ';' expected.
node_modules/typescript/lib/lib.core.d.ts(1168,41): error TS1005: '(' expected.
node_modules/typescript/lib/lib.core.d.ts(1169,26): error TS1005: '(' expected.
node_modules/typescript/lib/lib.core.d.ts(1170,1): error TS1128: Declaration or statement expected.
node_modules/typescript/lib/lib.core.es6.d.ts(1168,28): error TS1005: ';' expected.
node_modules/typescript/lib/lib.core.es6.d.ts(1168,31): error TS1005: ';' expected.
node_modules/typescript/lib/lib.core.es6.d.ts(1168,41): error TS1005: '(' expected.
node_modules/typescript/lib/lib.core.es6.d.ts(1169,26): error TS1005: '(' expected.
node_modules/typescript/lib/lib.core.es6.d.ts(1170,1): error TS1128: Declaration or statement expected.
node_modules/typescript/lib/lib.d.ts(1168,28): error TS1005: ';' expected.
node_modules/typescript/lib/lib.d.ts(1168,31): error TS1005: ';' expected.
node_modules/typescript/lib/lib.d.ts(1168,41): error TS1005: '(' expected.
node_modules/typescript/lib/lib.d.ts(1169,26): error TS1005: '(' expected.
node_modules/typescript/lib/lib.d.ts(1170,1): error TS1128: Declaration or statement expected.
node_modules/typescript/lib/lib.es6.d.ts(1168,28): error TS1005: ';' expected.
node_modules/typescript/lib/lib.es6.d.ts(1168,31): error TS1005: ';' expected.
node_modules/typescript/lib/lib.es6.d.ts(1168,41): error TS1005: '(' expected.
node_modules/typescript/lib/lib.es6.d.ts(1169,26): error TS1005: '(' expected.
node_modules/typescript/lib/lib.es6.d.ts(1170,1): error TS1128: Declaration or statement expected.

**Comment by bpasero** [comment]
Sorry, I was running with a globally installed 1.5.x and this caused all these issues. It works fine now, especially since I discovered https://github.com/Microsoft/TypeScript/wiki/Typings-for-npm-packages

I really like how this was pushed for 1.6, makings typings a lot more understandable!

**Comment by mhegazy** [comment]
glad that was resolved.
